### ISSUE TITLE
Enumeration option cannot be selected in exported variable of type enum

### ISSUE THREAD

**Issue by Torguen** [godotengine/godot#96368]
### Tested versions

v4.4.dev1.official [28a72fa43]

### System information

w10 64

### Issue description

Note this code:
Father.gd has only one enumeration "TestEnum"

![Captura2](https://github.com/user-attachments/assets/177898f1-77c9-4515-bd48-285be1e366f5)

test.gd inherits from Father.gd
test.gd declares a new enumeration "TestEnum2"

![Captura](https://github.com/user-attachments/assets/4a354e7b-5d12-4ff2-84fc-7d72eaa1c737)

If you initialize the enumeration so that the values ​​do not match the enumeration in Father.gd, the option in the enumeration "Test3" in test.gd will not be selectable in the editor.

![Captura](https://github.com/user-attachments/assets/923f3ebc-29ed-4921-9446-08e1588f4f07)

Note the comments above each enumeration in "test.gd"

### Steps to reproduce

Try selecting the "Test3" option in the enumeration in the editor

### Minimal reproduction project (MRP)

[testenum.zip](https://github.com/user-attachments/files/16823826/testenum.zip)

**Comment by AThousandShips** [comment]
This is because it's identical to `TEST`, what you actually have here is:
```gdscript
enum TestEnum2 { TEST = 2, TEST2 = 1, TEST3 = 2 }
```
Enums with no explicit value will take the last declared value plus one, see [here](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#enums)

This is something to potentially add a warning for in the editor, but otherwise nothing is broken or wrong here

**Comment by HolonProduction** [comment]
I don't read anything about this behavior in the documentation you linked, in the examples the values are always defined in order so it can't be interfered either.

So this should be documented, but to me it feels like an oversight, it should probably be the max of all existing values + 1 and not the last declared value + 1.

From a usability standpoint we definitely don't want to automatically add a conflicting value.

**Comment by AThousandShips** [comment]
That would massively break compatibility and be unexpected from pretty much every other programming language I've seen, so changing how the values are assigned is not appropriate IMO, at least not until far far later

It does show that the values are assigned the previous value plus one, but should be made a bit more clear, but it doesn't imply anything about it *not* assigning the previous value plus one or that it makes a unique value

**Comment by HolonProduction** [comment]
> It does show that the values are assigned the previous value plus one, but should be made a bit more clear, but it doesn't imply anything about it _not_ assigning the previous value plus one or that it makes a unique value

But that is what users expect from enums and while not confirming it, the documentation doesn't refute this assumption either.

> but do you have any examples of a language

To be honest I'm mostly involved with languages that use the concept of enum classes were this isn't really a topic (e.g. dart, kotlin, python), but python for one uses the max value + 1 approach for its `auto` method (https://playground.programiz.com/?share_id=8m6uLExKIzTNF)

But I think the fact, that a lot of newer languages opt to implement enums in this way is a strong hint about what users use enums for in high level languages, a unique value.

> Expected:
> 
>     * `BAZ_1 == 6, BAZ_2 == 6`
> 
> 
> Actual:
> 
>     * `BAZ_1 == 11, BAZ_2 == 6`

Why would it be expected that members of different enums are identical? I think the assumption that enum members are unique is much more wide spread. If comparing values from different enums is something that code relies on, the value should be made explicit for all values IMO.

> Note that this would break _existing_ code

Totally fair, we shouldn't change the behavior I just noticed that it isn't what I would expect, so I think the docs should clarify it.

But in my opinion we should add an error for the case that godot does automatically assign a conflicting value, if it is intended it should be made explicit. The error could be set to warning in the settings for old code that relies on the behavior, but for most GDScript users this is behavior is just a foot-gun.

**Comment by AThousandShips** [comment]
> But that is what users expect from enums and while not confirming it

What do you base this on?

> Why would it be expected that members of different enums are identica?

Because they're declared in exactly the same way *except* an unrelated earlier value is different 

> I think the assumption that enum members are unique is much more wide spread

Given that C++, C, and C# behaves this way I'd doubt that unless most users have little other programming experience

> But in my opinion we should add an error for the case that godot does automatically assign a conflicting value

That warning would be annoying generally as those who assign a non-linear value would largely know what they're doing, and most users just don't assign explicit values I'd suspect

So far I've never seen a bug report or similar of someone being confused by this, except if that's the cause of the confusion in this report, it seems to be more that the editor doesn't show a duplicate being selected

**Comment by HolonProduction** [comment]
> What do you base this on?

This issue and my own assumptions, admittedly not a solid data foundation :sweat_smile:, but I don't know were I would get better data on this.

> I'd doubt that unless most users have little other programming experience

I always felt like newcomers and inexperienced programmers were a large target audience for GDScript.

> So far I've never seen a bug report or similar of someone being confused by this

I don't have either, but I also never really have seen someone assigning explicit values out of order, which is kind of required to run into this.

> expect most people to make the assumption that it assigns the largest value plus one ever unless it's the same as last value plus one

That isn't the assumption that I think that users have. I think users don't care about enum values without explicit values at all. Personally I wouldn't have any expectations about the actual value of such a thing, if I don't add an explicit value I won't use or rely on the value at all, I just want an unique identifier in this case. But yeah this lacks a real data basis again.

**Comment by Zireael07** [comment]
> I always felt like newcomers and inexperienced programmers were a large target audience for GDScript.

Same here

**Comment by AThousandShips** [comment]
> I always felt like newcomers and inexperienced programmers were a large target audience for GDScript.

Absolutely, so we can't really expect *any* specific assumptions they make, regardless, but from my experience I haven't seen people making assumptions that confuse them with how enums are initialized

Will write up a documentation improvement to clarify further

**Comment by HolonProduction** [comment]
> so we can't really expect _any_ specific assumptions they make

Which would speak for an error or warning (but I find that newcomers don't really find warnings since they aren't inlined)

But I guess my assumptions are definitely biased towards enum classes so I'll leave it at that. And an updated documentation goes a long way already :+1:

**Comment by AThousandShips** [comment]
I think a warning or even an error when *exporting* enums with duplicate values would be useful as it *will* make the inspector behave weirdly (and we can't fix that without making it *more* confusing, as enums are stored by value, not by key, and it would be *more* wrong to show `Test 3` selected instead of `Test 1`)

But if reasonably many users are having issues with accidental duplicate enum values that's something to look at for a general warning, but would need some data on that (and would suggest opening a proposal for this instead, this issue should probably focus on the editor interface going forward)

**Comment by AThousandShips** [comment]
Doing some testing on a solution for reducing the confusion in the editor by disabling (and possibly marking as "duplicate") duplicate items so you can't select them, might simply not add them as well but will test different solutions to make this less confusing in the editor

**Comment by kleonc** [comment]
Isn't this issue already resolved by https://github.com/godotengine/godot-docs/pull/9872 and #96386? 🤔

Adding potential warning is tracked separately in https://github.com/godotengine/godot-proposals/issues/10618.

---

I guess what could help further to avoid confusion would be showing actual integer values for the given entries in the tooltip, currently such values are not shown. I've opened #102734 addressing this.
